### PR TITLE
wfst-based decoding

### PR DESCRIPTION
**Pull request by xu-gaopeng**
This PR add WFST-based decoding to E2E models
Ref:https://github.com/k2-fsa/k2 
1.add functions for dealing with WFST
 2.integrate K2 create token/lexicon/grammar fst, composing decoding graph 
3.add scrips to make graph (to do)

### CODE REVIEW

**Comment by sw005320** [comment]
Thanks!
This is a really cool function!
I have the following high-level comments.

- Do you also have some working recipe examples?
- We also want to use it for espnet2 as well.

**Comment by xu-gaopeng** [comment]
> Thanks!
> This is a really cool function!
> I have the following high-level comments.
> 
> * Do you also have some working recipe examples?
> * We also want to use it for espnet2 as well.

Yes, I will finish it in the near future

File: espnet/asr/pytorch_backend/asr.py [hunk 1]
```diff
@@ -1116,6 +1116,13 @@ def recog(args):
                                 nbest_hyps[n]["score"] += hyps[n]["score"]
                 elif hasattr(model, "is_rnnt"):
                     nbest_hyps = model.recognize(feat, beam_search_transducer)
+                elif hasattr(model, "decoder_mode") and args.wfst_decode:
+                    grpha = torch.load('data/graph/HLG.pt')
```
Comment (sw005320) [inline espnet/asr/pytorch_backend/asr.py:1120]: It's better to specify the file path of `HLG.pt` instead of this hard-coded path.

File: espnet/nets/pytorch_backend/e2e_asr_transformer.py [hunk 2]
```diff
@@ -160,6 +162,44 @@ def reset_parameters(self, args):
         # initialize parameters
         initialize(self, args.transformer_init)
 
+    def wfst_based_decoding(self, x, HLG, recog_args, char_list=None):
```
Comment (sw005320) [inline espnet/nets/pytorch_backend/e2e_asr_transformer.py:165]: Instead of adding it under transformer, I think you can add it in https://github.com/espnet/espnet/blob/master/espnet/nets/pytorch_backend/e2e_asr.py
Then, you can use it for all our architectures.

Probably you can move this method in addition to 
https://github.com/espnet/espnet/blob/c744073fc66328d27642f0083514a3a3625c677b/espnet/nets/pytorch_backend/e2e_asr.py#L379-L399

@ShigekiKarita and @YosukeHiguchi, could you give us some comments?

**Review by sw005320 — COMMENTED** [review]
I leave some comments.
I don't want to make a PR bigger, so you don't have to do the following items in this PR, but it would be great if you consider them later PRs.

- [ ] add a unit test
- [ ] add a recipe

File: espnet/nets/pytorch_backend/e2e_asr_transformer.py (hunk 2 above)
Comment (sw005320) [reply espnet/nets/pytorch_backend/e2e_asr_transformer.py:165]: @YosukeHiguchi, do you think we should make it similar to your mask CTC inference, or my above idea would be OK?

File: espnet/nets/pytorch_backend/e2e_asr_transformer.py [hunk 3]
```diff
@@ -160,6 +162,44 @@ def reset_parameters(self, args):
...
+            encoder_out)
+        log_probs[:, :, 0] += -1
+        dense_fsa_vec_seg = torch.stack(
+             (torch.tensor(range(encoder_out_lens.shape[0])),
+             torch.zeros(encoder_out_lens.shape[0]),
+             encoder_out_lens.to("cpu")), 1).to(torch.int32)
+        indices = torch.argsort(dense_fsa_vec_seg[:, 2], descending=True)
+        dense_fsa_vec_seg = dense_fsa_vec_seg[indices]
+        dense_fsa_vec = k2.DenseFsaVec(log_probs, dense_fsa_vec_seg)
+        lattices = k2.intersect_dense_pruned(HLG, dense_fsa_vec, 30.0, 7.0, recog_args.beam_size, 10000)
```
Comment (csukuangfj) [inline espnet/nets/pytorch_backend/e2e_asr_transformer.py:186]: According to the documentation of `k2.intersect_dense_pruned`
https://k2.readthedocs.io/en/latest/python_api/fsa_algo.html#k2.intersect_dense_pruned

The position where 7.0 occupies is the `beam_size` (i.e., `output_beam`). The current position for
`recog_args.beam_size` is actually `min_active_states`.

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: espnet/nets/pytorch_backend/e2e_asr_transformer.py (hunk 2 above)
Comment (YosukeHiguchi) [reply espnet/nets/pytorch_backend/e2e_asr_transformer.py:165]: If my understanding is correct, E2E classes in e2e_asr.py and e2e_asr_transformer.py are defined independently of each other. So this method should be added in both classes so that RNN-based models from e2e_asr.py and Transformer-based models from e2e_asr_transformer.py can use it.

**Comment by ShigekiKarita** [comment]
Wow, this is cool. Can you add new tests that check your funcs can run with toy data (e.g. small tensors) to keep compatibility with k2? If k2 installation is difficult in our CI script, you can skip that now (see kenlm test: https://github.com/espnet/espnet/blob/36b62ae48f9605fe6f05bb59fd0b6981126bba41/test/test_ngram.py#L6

**Comment by sw005320** [comment]
@xu-gaopeng, can you revisit this PR?
I think this is a very good function.

**Comment by danpovey** [comment]
If he does not respond, one of us can help..

**Comment by sw005320** [comment]
> If he does not respond, one of us can help..

OK, that sounds great.

**Comment by xu-gaopeng** [comment]
Due to the delay of personal affairs, I will focus on solving related problems in recent weeks.

**Comment by sw005320** [comment]
> Due to the delay of personal affairs, I will focus on solving related problems in recent weeks.

I see. No problem.

@danpovey, can I ask your team's help?
I think we can follow the basic implementation based on this PR, but adding a recipe would be important.
I'll help with this part.

**Comment by danpovey** [comment]
OK, someone will work on it.

**Comment by danpovey** [comment]
@sw005320: our team member @glynpu will work on this as he has been working the most with ESPNet.
We may also ask for your help in figuring out why our WER results are lagging behind those of ESPNet.

**Comment by sw005320** [comment]
> @sw005320: our team member @glynpu will work on this as he has been working the most with ESPNet.

Sounds good!

> We may also ask for your help in figuring out why our WER results are lagging behind those of ESPNet.

Yes, as I mentioned, I'm happy to help.
If you have some related discussions on the issue or PR, please invite me (or email or whatever, anything is fine).

**Comment by glynpu** [comment]
> * We also want to use it for espnet2 as well.

Maybe I could try do this task in espnet2 first.

In the past weeks, I do some analysis by combing k2 and the SOTA espnet2 released model, i.e. https://zenodo.org/record/4604066#.YNmZKugzaUk 
Results are shown below(Maybe the result could be better if tuned):

  | encoder + k2-ctc-decode without rescore | encoder + k2 ctc decode + 6-layer decoder nbest rescore | encoder + k2-ctc-decode + transformer lm nbest rescore | encoder + k2 ctc decode + 6-layer-decoder nbest rescore + transformer lm nbest rescore
-- | -- | -- | -- | --
test-clean wer % number_paths = 100 | 2.97 | 2.64 | 2.43 | 2.35
decoder algorithm in espnet | * | * | * | 2.1% decoder algorithm in espnet

Note: 
1. k2-ctc-decode is similar to [this](https://github.com/k2-fsa/snowfall/blob/master/egs/librispeech/asr/simple_v1/bpe_ctc_att_conformer_decode.py#L67):
```
        with torch.no_grad():
            dense_fsa_vec = k2.DenseFsaVec(nnet_output, supervision_segments)

            lattices = k2.intersect_dense_pruned(ctc_topo, dense_fsa_vec, 20.0,
                                                 output_beam_size, 30, 10000)

        best_paths = k2.shortest_path(lattices, use_double_scores=True)
        hyps = get_texts(best_paths, indices)
```
2. rescore with number_paths=100 means that 100 seqs are extracted from lattices then they rescored with decoder or/and transformer lm. 

>*Yes, as I mentioned, I'm happy to help.
If you have some related discussions on the issue or PR, please invite me (or email or whatever, anything is fine).

[This pr](https://github.com/k2-fsa/snowfall/pull/219) is where we are discussing about reproducing a comparable result to previously mentioned espnet released model.
A known difference is that espnet use mean and variance normalization of fbank while snowfall don't. 
Does this explain the difference between snowfall's 3.32% and espnet's 2.97% on test-clean? @sw005320

**Comment by pzelasko** [comment]
I just noticed this PR and discussion :) @glynpu if you want to check snowfall with MVN, you may use this class from Lhotse either as a part of the network or in transforms inside the dataloader.

https://github.com/lhotse-speech/lhotse/blob/master/lhotse/dataset/signal_transforms.py#L17

**Comment by sw005320** [comment]
@glynpu, that's a very impressive result.
Thanks for the report.
Did you increase the number of n-best sequences?
I just want to confirm whether the result is converged or not.

**Comment by sw005320** [comment]
> > * We also want to use it for espnet2 as well.
> 
> Maybe I could try to do this task in espnet2 first.

Can you do it when you have time?

**Comment by glynpu** [comment]
> Can you do it when you have time?

Sure.  
The result I posted in previous comment is obtained by a rough solution. 
In the past week, [an optimized decoding pipeline](https://github.com/k2-fsa/snowfall/pull/227) is implemented, which is about encoder + (ctc + 3-gram) + 4-gram lattice rescore + nbest_rescore with transformer decoder net.

I am going to port [this optimized decoding pipeline](https://github.com/k2-fsa/snowfall/pull/227) to espnet once it's is reviewed and approved.
Could you please check whether it is suitable for espnet? I am glad to make any modifications to make it easily been integrated into espnet.

**Comment by glynpu** [comment]
>Could you please check whether this decoding method suitable for espnet? I am glad to make any modifications to make it easily been integrated into espnet.

By suitable, wht I mean is:
 in espnet,  transformer decoder and neural language model(full-scorers) is used as first-step decoding then ctc is used as a rescoring part(partial-scores).

To use k2, The whole decoding part may use ctc as first-step decoding, then n-gram language model or transfromer-decoder    or neural net language model is used as a rescore part.

**Comment by sw005320** [comment]
> > Could you please check whether this decoding method suitable for espnet? I am glad to make any modifications to make it easily been integrated into espnet.
> 
> By suitable, wht I mean is:
> in espnet, transformer decoder and neural language model(full-scorers) is used as first-step decoding then ctc is used as a rescoring part(partial-scores).
> 
> To use k2, The whole decoding part may use ctc as first-step decoding, then n-gram language model or transfromer-decoder or neural net language model is used as a rescore part.

Yes, I see the point.
This is exactly what we want.
@takaaki-hori (as you may know, the search expert) will also help you.

**Comment by glynpu** [comment]
>Yes, I see the point.
This is exactly what we want.
@takaaki-hori (as you may know, the search expert) will also help you.

Thanks! [This implementation](https://github.com/k2-fsa/snowfall/pull/227) will be integrated into espnet after some revisions, which won't take too much time.

File: espnet/nets/pytorch_backend/e2e_asr_transformer.py [hunk 4]
```diff
@@ -160,6 +162,44 @@ def reset_parameters(self, args):
...
+        :return: decoding result
+        :rtype: list
+        """
+        self.eval()
+        encoder_out, encoder_mask = self.encoder(x, None)
+        log_probs = self.ctc.log_softmax(
+            encoder_out)
+        log_probs[:, :, 0] += -1
+        dense_fsa_vec_seg = torch.stack(
+             (torch.tensor(range(encoder_out_lens.shape[0])),
```
Comment (takaaki-hori) [inline espnet/nets/pytorch_backend/e2e_asr_transformer.py:180]: Where is `encoder_out_lens` defined? I don't see the line where `encoder_out_lens` is defined.

File: espnet/nets/pytorch_backend/e2e_asr_transformer.py [hunk 5]
```diff
@@ -160,6 +162,44 @@ def reset_parameters(self, args):
...
+        labels = k2.RaggedInt(shape, labels.values())
+        assert (labels.num_axes() == 2)
+        ans = torch.zeros(indices.shape, device=indices.device, dtype=torch.long)
+        ans[indices] = torch.arange(0, indices.shape[0], device=indices.device)
+        labels, _ = k2.ragged.index(labels,
+                        ans.to(dtype=torch.int32,
+                        device=best_paths.device))
+        hyps = k2.ragged.to_list(labels)
+        
+        return hyps
```
Comment (takaaki-hori) [inline espnet/nets/pytorch_backend/e2e_asr_transformer.py:201]: I am a bit concerned about the return type. ESPnet expects that the return type here is something like a list of dict, where each dict includes keys `yseq` and `score`, and the value of `yseq` is an array of token ids. Is the return type designed like this? Did you check if the current implementation runs without any errors?

File: egs/aishell/asr1/local/fst/prepare_fst.sh [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    cp $dict data/fst/units.txt
+    local/fst/prepare_dict.py $dict ${data}/resource_aishell/lexicon.txt \
+        data/fst/lexicon.txt
+    echo "<unk> <unk>" >>data/fst/lexicon.txt
+    local/fst/make_L.sh data/fst data/fst/tmp data/fst/lang 
+   
+    # Train lm
+    lm=data/local/lm
+    mkdir -p $lm
+    text2token.py -s 1 -n 1 data/train/text | cut -f 2- -d" " > $lm/text
```
Comment (mlecauchois) [inline egs/aishell/asr1/local/fst/prepare_fst.sh:24]: Just curious, is there a reason to produce a char-based LM instead of a word-based one ? (default of `text2token.py` is char-based split)
